core/src/subgraph/context/instance.rs

```diff
@@ -204,11 +202,121 @@ where
         }
     }
 
-    pub fn hosts(&self) -> &[Arc<T::Host>] {
-        &self.hosts
+    /// Returns all hosts which match the trigger's address.
+    /// This is a performance optimization to reduce the number of calls to `match_and_decode`.
+    pub fn hosts_for_trigger(
+        &self,
+        trigger: &TriggerData<C>,
+    ) -> Box<dyn Iterator<Item = &T::Host> + Send + '_> {
+        self.hosts.iter_by_address(trigger.address_match())
     }
 
     pub(super) fn causality_region_next_value(&mut self) -> CausalityRegion {
         self.causality_region_seq.next_val()
     }
+
+    #[cfg(debug_assertions)]
+    pub fn hosts(&self) -> &[Arc<T::Host>] {
+        &self.hosts.hosts()
+    }
+}
+
+/// Runtime hosts, one for each data source mapping.
+///
+/// The runtime hosts are created and added to the vec in the same order the data sources appear in
+/// the subgraph manifest. Incoming block stream events are processed by the mappings in this same
+/// order.
+///
+/// This structure also maintains a partition of the hosts by address, for faster trigger matching.
+/// This partition uses the host's index in the main vec, to maintain the correct ordering.
+struct Hosts<C: Blockchain, T: RuntimeHostBuilder<C>> {
+    hosts: Vec<Arc<T::Host>>,
+
+    // The `usize` is the index of the host in `hosts`.
+    hosts_by_address: HashMap<Box<[u8]>, Vec<usize>>,
+    hosts_without_address: Vec<usize>,
+}
+
+impl<C: Blockchain, T: RuntimeHostBuilder<C>> Hosts<C, T> {
+    fn new() -> Self {
+        Self {
+            hosts: Vec::new(),
+            hosts_by_address: HashMap::new(),
+            hosts_without_address: Vec::new(),
+        }
+    }
+
+    fn hosts(&self) -> &[Arc<T::Host>] {
+        &self.hosts
+    }
+
+    fn last(&self) -> Option<&Arc<T::Host>> {
+        self.hosts.last()
+    }
+
+    fn len(&self) -> usize {
+        self.hosts.len()
+    }
+
+    fn push(&mut self, host: Arc<T::Host>) {
+        self.hosts.push(host.cheap_clone());
+        let idx = self.hosts.len() - 1;
+        let address = host.data_source().address();
+        match address {
+            Some(address) => {
+                self.hosts_by_address
+                    .entry(address.into())
+                    .or_default()
+                    .push(idx);
+            }
+            None => {
+                self.hosts_without_address.push(idx);
+            }
+        }
+    }
+
+    fn pop(&mut self) {
+        let Some(host) = self.hosts.pop() else { return };
+        let address = host.data_source().address();
+        match address {
+            Some(address) => {
+                // Unwrap and assert: The same host we just popped must be the last one in `hosts_by_address`.
+                let hosts = self.hosts_by_address.get_mut(address.as_slice()).unwrap();
+                let idx = hosts.pop().unwrap();
+                assert_eq!(idx, self.hosts.len());
```

Comment:
Is this running inside some panic handler that will stop the subgraph if this assert gets tripped? Otherwise, it would be better to return an error so that only this subgraph stops, and others can continue

Reply:
Thanks for the review, I've checked and this panic will only stop the runner thread for the subgraph that trips it.